explorer/interpreter/type_checker.cpp

```diff
@@ -4401,6 +4426,83 @@ auto TypeChecker::TypeCheckGenericBinding(GenericBinding& binding,
   return Success();
 }
 
+auto GetSize(Nonnull<const Value*> from) -> size_t {
+  switch (from->kind()) {
+    case Value::Kind::TupleType:
+    case Value::Kind::TupleValue: {
+      const auto& from_tup = cast<TupleValueBase>(*from);
+      CARBON_CHECK(!from_tup.elements().empty());
+      return from_tup.elements().size();
+    }
+    case Value::Kind::StaticArrayType: {
+      const auto& from_arr = cast<StaticArrayType>(*from);
+      CARBON_CHECK(from_arr.has_size());
+      return from_arr.size().value();
+    }
+    default:
+      return 0;
+  }
+}
+
+auto TypeChecker::DeduceType(Nonnull<const Value*> type,
+                             Nonnull<const Value*> expected)
+    -> Nonnull<const Value*> {
+  switch (type->kind()) {
+    case Value::Kind::IntValue:
+    case Value::Kind::FunctionValue:
+    case Value::Kind::DestructorValue:
+    case Value::Kind::BoundMethodValue:
+    case Value::Kind::PointerValue:
+    case Value::Kind::LocationValue:
+    case Value::Kind::BoolValue:
+    case Value::Kind::TupleValue:
+    case Value::Kind::StructValue:
+    case Value::Kind::NominalClassValue:
+    case Value::Kind::AlternativeValue:
+    case Value::Kind::BindingPlaceholderValue:
+    case Value::Kind::AddrValue:
+    case Value::Kind::AlternativeConstructorValue:
+    case Value::Kind::StringValue:
+    case Value::Kind::UninitializedValue:
+    case Value::Kind::ImplWitness:
+    case Value::Kind::BindingWitness:
+    case Value::Kind::ConstraintWitness:
+    case Value::Kind::ConstraintImplWitness:
+    case Value::Kind::ParameterizedEntityName:
+    case Value::Kind::MemberName:
+    case Value::Kind::IntType:
+    case Value::Kind::BoolType:
+    case Value::Kind::TypeType:
+    case Value::Kind::PointerType:
+    case Value::Kind::FunctionType:
+    case Value::Kind::StructType:
+    case Value::Kind::TupleType:
+    case Value::Kind::NominalClassType:
+    case Value::Kind::InterfaceType:
+    case Value::Kind::NamedConstraintType:
+    case Value::Kind::ConstraintType:
+    case Value::Kind::ChoiceType:
+    case Value::Kind::VariableType:
+    case Value::Kind::StringType:
+    case Value::Kind::AutoType:
+    case Value::Kind::TypeOfParameterizedEntityName:
+    case Value::Kind::TypeOfMemberName:
+    case Value::Kind::TypeOfMixinPseudoType:
+    case Value::Kind::TypeOfNamespaceName:
+    case Value::Kind::AssociatedConstant:
+    case Value::Kind::MixinPseudoType: {
+      return expected;
+    }
+    case Value::Kind::StaticArrayType: {
+      const auto& arr = cast<StaticArrayType>(*type);
+      CARBON_CHECK(!arr.has_size());
+      const size_t size = GetSize(expected);
+      CARBON_CHECK(size != 0);
```

Comment:
After I've fixed the 0-sized arrays(not pushed yet) this particular test fails with error `use of uninitialized variable a`, `var a: [i32;0] = ();` fixes the test, but for me it seems ok for zero-length array to be "uninitialized", because we actually have nothing to initialize. What do you think?

Btw, any way currently in Carbon to get the size of an array? Anything like `b.size()` or `size(b)`?

Reply:
I think there's a separate issue around array initialization. #2816 did some changes here recently -- if `var a: [i32;0]` leaves `a` considered uninitialized, that'd indicate the fix didn't fully address the issue (#2812).

I think there's currently no way to get the size of an array -- feature not yet designed. I think `b.size()` would probably work and be preferred.